### PR TITLE
CFE-555: Add additional fields to configure DNS management on ingresscontrollers

### PR DESCRIPTION
**Pull request by thejasn**
Add additional fields to configure DNS management on ingresscontrollers
 ---
 This PR, introduces 2 fields to the IngressController API,
 - `DNSManagementPolicy` of type `LoadBalancerDNSManagementPolicy` to the ingresscontroller spec. 
 - `DNSManagementPolicy` to the DNSRecord spec.
 
 
RFE: https://issues.redhat.com/browse/RFE-895
EP: https://github.com/openshift/enhancements/pull/1114

### CODE REVIEW

File: operator/v1/types_ingress.go [hunk 1]
```diff
@@ -912,7 +936,7 @@ type SyslogLoggingDestinationParameters struct {
 
 // ContainerLoggingDestinationParameters describes parameters for the Container
 // logging destination type.
-type ContainerLoggingDestinationParameters struct {
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go]: Can we remove these white space changes, they don't seem relevant to the PR

File: operatoringress/v1/types.go [hunk 2]
```diff
@@ -115,6 +123,21 @@ const (
 	ARecordType DNSRecordType = "A"
 )
 
+// DNSManagementPolicy is a policy for configuring how the dns controller
+// manages DNSRecords.
+//
+// +kubebuilder:validation:Enum=Managed;Unmanaged
```
Comment (JoelSpeed) [inline operatoringress/v1/types.go:129]: Because you've specified this here, you don't need to specify it each time you use the enum. If you look at the generated CRD, you have the enum twice because of this

**Review by JoelSpeed — COMMENTED** [review]
Just to confirm, this is a new field which is being added to an existing, already released API?

What happens to existing users/clients? The API will default the field to managed, will old clients try to force the field to be dropped because they are unaware of the new field? 
Just wondering if there's a way that a user sets this to `Unmanaged`, then a controller that doesn't new about the new field clears the field, causing it to be defaulted back to `Managed`.

Also, do you intend to have any rules about switching between the two? Is it ok for the operator to take on management of the resource once it is unmanaged?

File: operator/v1/types_ingress.go [hunk 3]
```diff
@@ -384,8 +384,32 @@ type LoadBalancerStrategy struct {
 	//
 	// +optional
 	ProviderParameters *ProviderLoadBalancerParameters `json:"providerParameters,omitempty"`
+
+	// dnsManagementPolicy indicates if the lifecyle of the wildcard DNS record
+	// associated with the load balancer service will be managed by
+	// the ingress operator.
+	//
+	// +kubebuilder:default:="Managed"
```
Comment (deads2k) [inline operator/v1/types_ingress.go:392]: Before this field was added, was the default logically Managed?

File: operatoringress/v1/types.go [hunk 4]
```diff
@@ -55,6 +55,14 @@ type DNSRecordSpec struct {
...
 	// +required
 	RecordTTL int64 `json:"recordTTL"`
+	// dnsManagementPolicy denotes the current policy applied on the dns record.
+	// Records that have policy set as "Unmanaged" are ignored by the ingress
+	// operator and should be managed at the discretion of the cluster admin.
+	//
+	// +kubebuilder:default:="Managed"
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=Managed;Unmanaged
+	DNSManagementPolicy DNSManagementPolicy `json:"dnsManagementPolicy"`
```
Comment (deads2k) [inline operatoringress/v1/types.go:65]: Why would we have a DNSRecord resource in cases where we do not managed the DNS?

File: operatoringress/v1/types.go [hunk 5]
```diff
@@ -55,6 +55,14 @@ type DNSRecordSpec struct {
 	// +kubebuilder:validation:Minimum=0
 	// +required
 	RecordTTL int64 `json:"recordTTL"`
+	// dnsManagementPolicy denotes the current policy applied on the dns record.
+	// Records that have policy set as "Unmanaged" are ignored by the ingress
+	// operator and should be managed at the discretion of the cluster admin.
```
Comment (deads2k) [inline operatoringress/v1/types.go:60]: The doc on this type explicitly says: 

> Cluster admin manipulation of this resource is not supported.

Why are we telling them to manage a thing based on this?

File: operator/v1/types_ingress.go (hunk 3 above)
Comment (thejasn) [reply operator/v1/types_ingress.go:392]: Yes, prior to this new field, the behavior is synonymous to `Managed`.

File: operatoringress/v1/types.go (hunk 4 above)
Comment (thejasn) [reply operatoringress/v1/types.go:65]: New created ingresscontrollers with `dnsManagementPolicy` set to `Unmanaged` will not create a *DNSRecord*, but in scenarios where we update the IC from `Managed` (current default) to `Unmanaged` we will have an associated *DNSRecord* with `dnsManagementPolicy` set to `Unmanaged`. The cluster admin can later on clean up the resource as needed.

File: operatoringress/v1/types.go (hunk 5 above)
Comment (thejasn) [reply operatoringress/v1/types.go:60]: Basically there is no management that the cluster admins can do other than to delete the *DNSRecord* because once `dnsManagementPolicy` is set to `Unmanaged` the controller will ignore the resource silently and not perform any actions on it until the IC is updated back to `Managed` (will update the docs on this type). 

Maybe instead of telling
`should be managed at the discretion of the cluster admin`,
we could say
`the resource can be deleted at the discretion of the cluster admin`. 
WDYT?

**Comment by thejasn** [comment]
Hey @JoelSpeed, 
> Just to confirm, this is a new field which is being added to an existing, already released API?
> 
Yes, adding a new field to an existing API. 
> What happens to existing users/clients? The API will default the field to managed, will old clients try to force the field to be dropped because they are unaware of the new field? Just wondering if there's a way that a user sets this to `Unmanaged`, then a controller that doesn't new about the new field clears the field, causing it to be defaulted back to `Managed`.
> 
The current default behavior is the same as that of `Managed`, where the ingresscontroller implicitly manages the DNS. So on existing clients the field will be dropped but will not affect the current functionality of the ingresscontroller.

> Also, do you intend to have any rules about switching between the two? Is it ok for the operator to take on management of the resource once it is unmanaged?

No, currently there are no checks to restrict going from `Unmanaged` to `Managed` or vice versa. This scenario is supported, where it creates the required resources if it doesn't exist.

**Comment by JoelSpeed** [comment]
> No, currently there are no checks to restrict going from Unmanaged to Managed or vice versa. This scenario is supported, where it creates the required resources if it doesn't exist.

In other APIs where we have the idea of being unmanaged, typically that is so that a user can either add something that's required as a placeholder (so something else can consume it) or so that they can start making changes that are otherwise unsupported. Typically we don't let them go back to managed because we don't know what they did in the mean time and our controllers aren't going to be able to resolve that.

**Comment by thejasn** [comment]
> > No, currently there are no checks to restrict going from Unmanaged to Managed or vice versa. This scenario is supported, where it creates the required resources if it doesn't exist.
> 
> In other APIs where we have the idea of being unmanaged, typically that is so that a user can either add something that's required as a placeholder (so something else can consume it) or so that they can start making changes that are otherwise unsupported. Typically we don't let them go back to managed because we don't know what they did in the mean time and our controllers aren't going to be able to resolve that.
> 
> Just wondering what the use case for unmanaged is here and if we can expand on that so that we can make sure it's safe to go from unmanaged to managed?

`Unmanaged` has 2 different meanings here,
- On the ingresscontroller it means the operator will not create any *DNSRecord* if it doesn't exist, or it means it has unlinked itself from the associated *DNSRecord* (when you go from `Managed` to `Unmanaged`).
- On the *DNSRecord* it just denotes its current state.

Basically there is no other management that the cluster admins can do other than to delete the DNSRecord because once `dnsManagementPolicy` is set to `Unmanaged` the controller will ignore the resource silently and not perform any actions on it until the IC is updated back to `Managed` (any changes done on the CR will be reverted if CR is present when moving from `Unmanaged` to `Managed`). The *DNSRecord* is not auto deleted, to help the admins with all the needed information in a single place so that it can be used to create a DNS record manually post which they can be deleted.

**Comment by deads2k** [comment]
I'd like to avoid resources that can sometimes be user managed and sometimes not be user managed.  How about a behavior like this

1. setting IngressController.dnsManagementPolicy=Unmanaged always produces a DNSRecord with a DNSRecord.spec.dnsManagementPolicy=Unmanaged.  If a user deletes the DNSRecord kube object, it reappears. 
2. DNSRecord.spec.dnsManagementPolicy=Unmanaged produces no mutation on the cloud resource DNS record, but we **do** continue to maintain status. DNSRecordStatus grows conditions which indicate whether it is able to read the status or not.  Ideally this is done by actually reading from the cloud provider API every five minutes or so.

this keeps the property that DNSRecord API objects are never user modifiable, but keeps a goal of "ingress controller stops managing the cloud provider DNS record".

**Comment by thejasn** [comment]
> I'd like to avoid resources that can sometimes be user managed and sometimes not be user managed. How about a behavior like this
> 
>     1. setting IngressController.dnsManagementPolicy=Unmanaged always produces a DNSRecord with a DNSRecord.spec.dnsManagementPolicy=Unmanaged.  If a user deletes the DNSRecord kube object, it reappears.
> 
>     2. DNSRecord.spec.dnsManagementPolicy=Unmanaged produces no mutation on the cloud resource DNS record, but we **do** continue to maintain status. DNSRecordStatus grows conditions which indicate whether it is able to read the status or not.  Ideally this is done by actually reading from the cloud provider API every five minutes or so.
> 
> 
> this keeps the property that DNSRecord API objects are never user modifiable, but keeps a goal of "ingress controller stops managing the cloud provider DNS record".

Going ahead with your suggestion, `Unmanaged` now means the DNS record on the cloud provider is not managed by the operator and the *DNSRecord* CR will be retained with `.spec.dnsManagementPolicy` set as `Unmanaged`. The same has been updated on the EP (https://github.com/openshift/enhancements/pull/1114/commits/d4ed7ed73dfaf6b735f886a705bbd8e904c47269). 

>DNSRecordStatus grows conditions which indicate whether it is able to read the status or not.  Ideally this is done by actually reading from the cloud provider API every five minutes or so.

By this, do you mean introduce a new field in `DNSRecordStatus` to denote the current status of the DNS record on the cloud provider and poll to update the current status?

**Comment by JoelSpeed** [comment]
Are we all agreed on the approach described here https://github.com/openshift/api/pull/1200#issuecomment-1144081676?

If I understand correctly and we are, and the API changes reflect that, then I have no further feedback on the API, so LGTM

**Comment by thejasn** [comment]
cc @Miciah

File: operatoringress/v1/types.go [hunk 6]
```diff
@@ -55,6 +58,17 @@ type DNSRecordSpec struct {
 	// +kubebuilder:validation:Minimum=0
 	// +required
 	RecordTTL int64 `json:"recordTTL"`
+	// dnsManagementPolicy denotes the current policy applied on the dns record.
+	// Records that have policy set as "Unmanaged" are ignored by the ingress
+	// operator and the DNS record on the cloud provider is not managed by the
```
Comment (deads2k) [inline operatoringress/v1/types.go:63]: When set to unmanaged, what happens to the Status?  Keeping stale status would be a problem.

Comment (thejasn) [reply operatoringress/v1/types.go:63]: If an ingresscontroller is created with `Unmanaged` policy, the `status` is kept empty, so there no question of this being stale. When you update the `dnsManagementPolicy` to `Unmanaged`, the `status` from the previous state is retained as is, the `status` kind of has a consolidated view of all the necessary  information that the user/admin needs to manually create DNS records externally and also to perform a clean up of the DNS zone and DNS records that were provisioned on the cloud (we do not delete the cloud resources when set to `Unmanaged`). This was kind of the main reason we decided to retain the previous `status`, without which it could potentially leave the cloud DNS resources orphaned. 

I'm not entirely sure we could call `status` as stale because it still holds valid information regarding the DNS records, etc that are allocated on the cloud (at least until manual deletion on the cloud).

Comment (deads2k) [reply operatoringress/v1/types.go:63]: > I'm not entirely sure we could call `status` as stale because it still holds valid information regarding the DNS records, etc that are allocated on the cloud (at least until manual deletion on the cloud).

This isn't the case. If you monitored to see what tags existed, it would be true.  Similarly if the conditions were updated, this would be true.  But near as I can tell looking at the API, only the ID will remain correct.  You cannot reasonably keep stale status.  If you don't observe the status, you don't even know if the ID still exists.

**Comment by deads2k** [comment]
/approve
/assign @Miciah 

The behavior describe is good. I'll leave final phrasing and lgtm to @Miciah

File: operatoringress/v1/types.go [hunk 7]
```diff
@@ -55,6 +58,18 @@ type DNSRecordSpec struct {
 	// +kubebuilder:validation:Minimum=0
 	// +required
 	RecordTTL int64 `json:"recordTTL"`
+	// dnsManagementPolicy denotes the current policy applied on the dns
+	// record. Records that have policy set as "Unmanaged" are ignored by
+	// the ingress operator and the DNS record on the cloud provider is not
+	// managed by the operator and the DNSRecord status will be updated to
+	// Unknown condition, since it is externally managed. This record on the
+	// cloud provider can be deleted at the discretion of the cluster admin.
+	// It defaults to Managed. Valid values are: Managed and Unmanaged.
```
Comment (Miciah) [inline operatoringress/v1/types.go:67]: Some wordsmithing suggestions:

* Use consistent capitalization: "dns record" → "DNS record".
* Reword to avoid the run-on sentence: "A and b and c" → "A.  This means b and c."  ("A, b, and c" would also be acceptable.)
* Adjust terminology for status conditions: "status will be updated to Unknown condition" → "the 'Published' status condition will be updated to 'Unknown' status".
* Account for the fact that the DNS record will not exist if the ingresscontroller was created with `dnsManagementState: Unmanaged`: "The record on the cloud provider" → "Any existing record on the cloud provider".

```suggestion
	// dnsManagementPolicy denotes the current policy applied on the DNS
	// record. Records that have policy set as "Unmanaged" are ignored by
	// the ingress operator.  This means that the DNS record on the cloud
	// provider is not managed by the operator, and the "Published" status
	// condition will be updated to "Unknown" status, since it is externally
	// managed. Any existing record on the cloud provider can be deleted at
	// the discretion of the cluster admin.
	//
	// This field defaults to Managed. Valid values are "Managed" and
	// "Unmanaged".
```

**Comment by Miciah** [comment]
Lovely jubbly!
/lgtm

**Comment by Amrita42** [comment]
docs-approved

File: operatoringress/v1/types.go (hunk 4 above)
Comment (gcs278) [reply operatoringress/v1/types.go:75]: @thejasn just to want to be clear, a new IC will indeed have a dnsRecord, right? This might be an update since you wrote this comment.

Comment (thejasn) [reply operatoringress/v1/types.go:75]: Yes, this was clarified ([here](https://github.com/openshift/api/pull/1200#issuecomment-1144896101)).
